packages/datadog-plugin-aws-sdk/src/helpers.js

```diff
@@ -0,0 +1,172 @@
+'use strict'
+
+const helpers = {
+  finish (span, err) {
+    if (err) {
+      span.setTag('error', err)
+      this.addRequestIdTag(span, err)
+    }
+
+    span.finish()
+  },
+
+  addResponseTags (span, response, serviceName, config) {
+    if (!span) return
+
+    if (response.request) {
+      this.addServicesTags(span, response, serviceName)
+    }
+
+    this.addRequestIdTag(span, response)
+    config.hooks.request(span, response)
```

Comment:
Given the call to `this.addRequestIdTag(span, err)` in the `finish` helper, it's unclear if `addResponseTags` is called for errors. If that's not the case, then the hook will never be called.

Reply:
There's a weird edge case where errors will, in certain library versions, sometimes append the `requestId` to the `response.error` object itself instead of the `response` object, I will try to dig up the reference to this in their docs, but this is basically a defensive check for that. the service specific response tagging will be called for errors as well, as the `complete` event will always return the response, and the response will have a `response.error` key defined in the case of an error [ts link](
https://github.com/aws/aws-sdk-js/blob/5b51b4e3297f0f1f13d66eaff0c2e8ad5473de49/lib/request.d.ts#L129).

Either way since the tagging has been refactored to just use the base/service class structure (described here: https://github.com/DataDog/dd-trace-js/pull/856#discussion_r384422783) I've just removed this `addRequestIdTag`method and hardcoded the edge case check into the `finish` method

```
  finish (span, err) {
    if (err) {
      span.setTag('error', err)

      if (err.requestId) {
        span.addTags({ 'aws.response.request_id': err.requestId })
      }
    }

    span.finish()
  }
```